doc/Variables.md

```diff
@@ -0,0 +1,144 @@
+[`fuse_core::Variable`](fuse_core/include/fuse_core/variable.h) base class defines the minimum interface required
+for all derived variable types.
+
+* `Derived::type() -> std::string`
+
+  All derived Variables must implement a `type()` method that returns the fully-qualified class name. This is of the
+  form `namespace::ClassName`. Under most circumstances, the base class implementation will return the correct class
+  name. However, templated Variable classes may need to override the `type()` method to return the desired name.
+
+* `Derived::size() -> size_t`
+
+  The derived Variable type must return the number of dimensions of the variable. This will likely just return a fixed
+  constant.
+
+* `Derived::data() -> double*`
+
+  The _value_ portion of the derived Variable must be accessible from a contiguous memory location of size
+  `derived.size() * sizeof(double)`.
+
+* `Derived::uuid() -> fuse_core::UUID`
+
+  Each derived class is required to return a unique ID to act as the identity of the variable. Some functions for
+  generating UUIDs are provided [here](fuse_core/include/fuse_core/uuid.h).
+
+* `Derived::print(std::ostream& stream)`
+
+  It's nice to print variable information during debugging. Each
+  derived Variable is required to implement a `print()` method, but the details of exactly what to print are left to
+  the derived Variable class designer. At a minimum, the variable `uuid()` is suggested.
+
+* `Derived::clone() -> fuse_core::Variable::UniquePtr`
+
+  All derived variables are required to implement a `clone()`
+  method. This should be implemented as `return Derived::make_unique(*this)`. Because this definition requires the use
+  of the derived type, a common implementation could not be provided in the base class.
+
+* `Derived::localParameterization() -> ceres::LocalParameterization*`
+
+  This is a complex topic on its own. See the
+  [Ceres documentation](http://ceres-solver.org/nnls_modeling.html#localparameterization) for an in-depth discussion
+  of "local parameterizations" and their uses. If the derived Variable type requires a local parameterization, this
+  method may be overridden to provide it.
+
+* `SMART_PTR_DEFINITIONS(Derived);`
+
+  It is highly recommended that all derived Variable types include `SMART_PTR_DEFINITIONS(Derived);` in the public
+  interface. This defines some common smart pointer aliases, such as `Derived::SharedPtr` and `Derived::UniquePtr`.
+
+Additional member properties and member functions may be added to the derived Variable class. These can only be used
+when an object is created with a known type. Despite this limitation, providing some syntax sugar is encouraged, as
+it can make working with the derived Variable objects more satisfying. For example, `fuse_variables` classes provide
+named accessors for the individual dimension values. This allows use of `var.y()` in lieu of `var.data()[1]`.
```

Comment:
Some general thoughts:

- Suggest just saying "Variable" instead of "Variable type". They start off interchangeable in the beginning, but then "Variable type" becomes dominant, but I think it's unnecessary and cumbersome.
- When talking about Variable value and Variable identity, suggest _dropping_ Variable completely within the relevant sections, and just referring to _value_ or _identity_ (more relevant for the lengthier _identity_ description).
- Variable capitalization of "Variable". Don't have a strong opinion on yay or nay, or whether to italicize, but it should be consistent so that it's clear when we're talking about a Fuse Variable.

Reply:
I think also taking one of the existing Variable types within fuse and walking through the reasoning behind it may help to illustrate the core concepts, especially if there is a chance to weigh pros and cons of the way it was defined. Again, an onerous task, but just leaving it here as a suggestion.